src/compiler/checker.ts

```diff
@@ -28040,6 +28036,33 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getUnionType(map(clauseWitnesses, text => text ? narrowTypeByTypeName(type, text) : neverType));
         }
 
+        function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
+            const clauses = switchStatement.caseBlock.clauses.slice(clauseStart, clauseEnd);
+            const clausesType = narrowTypeForTrueClauses(type, clauses);
+            const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);
+            const hasDefaultClause = clauseStart === clauseEnd || (defaultIndex >= clauseStart && defaultIndex < clauseEnd);
+            if (hasDefaultClause) {
+                // If we have a default in this set of clauses, then the type could also be any of the types
+                // that aren't covered by the other cases.
+
+                const clausesBefore = switchStatement.caseBlock.clauses.slice(0, clauseStart);
+                const clausesAfter = switchStatement.caseBlock.clauses.slice(clauseEnd);
+
+                const before = narrowTypeForTrueClauses(type, clausesBefore);
+                const after = narrowTypeForTrueClauses(type, clausesAfter);
+                const other = getUnionType([before, after]);
+
+                const typeNotOther = filterType(type, t => !isTypeSubsetOf(t, other));
+
+                return getUnionType([clausesType, typeNotOther]);

```

Comment:
I have no doubt that there's a more efficient way to do it, but I'm not sure how much it matters.